src/V3SchedUtil.cpp

```diff
@@ -116,64 +114,103 @@ static AstCFunc* splitCheckCreateNewSubFunc(AstCFunc* ofuncp) {
+                             const std::unordered_map<const AstVar*, AstVarScope*>& argVscps) {
+    FileLine* const flp = subFuncp->fileline();
+    AstCCall* const callp = new AstCCall{subFuncp->fileline(), subFuncp};
+    callp->dtypeSetVoid();
+    // Pass arguments through to subfunction
+    for (AstVar* argp = ofuncp->argsp(); argp; argp = VN_AS(argp->nextp(), Var)) {
+        UASSERT_OBJ(argp->direction() == VDirection::CONSTREF, argp, "Unexpected direction");
+        callp->addArgsp(new AstVarRef{flp, argVscps.at(argp), VAccess::READ});
+    }
+
+    bool containsAwait = false;
+    subFuncp->foreach([&](AstNodeExpr* exprp) {
+        // Record if it has a CAwait
+        if (VN_IS(exprp, CAwait)) containsAwait = true;
+        // Redirect references to arguments to the clone in the sub-function
+        if (AstVarRef* const refp = VN_CAST(exprp, VarRef)) {
+            if (AstVarScope* const vscp = VN_AS(refp->varp()->user3p(), VarScope)) {
+                refp->varp(vscp->varp());
+                refp->varScopep(vscp);
+            }
+        }
+    });
+
+    if (ofuncp->isCoroutine() && containsAwait) {  // Wrap call with co_await
+        subFuncp->rtnType("VlCoroutine");
+        AstCAwait* const awaitp = new AstCAwait{flp, callp};
+        awaitp->dtypeSetVoid();
+        ofuncp->addStmtsp(awaitp->makeStmt());
+    } else {
+        ofuncp->addStmtsp(callp->makeStmt());
+    }
+}
+
 // Split large function according to --output-split-cfuncs
-void splitCheck(AstCFunc* ofuncp) {
+void splitCheck(AstCFunc* const ofuncp) {
+    UASSERT_OBJ(!ofuncp->varsp(), ofuncp, "Can't split function with local variables");
     if (!v3Global.opt.outputSplitCFuncs() || !ofuncp->stmtsp()) return;
     if (ofuncp->nodeCount() < v3Global.opt.outputSplitCFuncs()) return;
 
-    int func_stmts = 0;
-    const bool is_ofuncp_coroutine = ofuncp->isCoroutine();
-    AstCFunc* funcp = nullptr;
-
-    const auto finishSubFuncp = [&](AstCFunc* subFuncp) {
-        ofuncp->scopep()->addBlocksp(subFuncp);
-        AstCCall* const callp = new AstCCall{subFuncp->fileline(), subFuncp};
-        callp->dtypeSetVoid();
-
-        if (is_ofuncp_coroutine && subFuncp->exists([](const AstCAwait*) {
-                return true;
-            })) {  // Wrap call with co_await
-            subFuncp->rtnType("VlCoroutine");
-
-            AstCAwait* const awaitp = new AstCAwait{subFuncp->fileline(), callp};
-            awaitp->dtypeSetVoid();
-            ofuncp->addStmtsp(awaitp->makeStmt());
-        } else {
-            ofuncp->addStmtsp(callp->makeStmt());
+    // Need to find the AstVarScopes for the function arguments. They should be in the same Scope.
+    std::unordered_map<const AstVar*, AstVarScope*> argVscps;
+    for (AstVar* argp = ofuncp->argsp(); argp; argp = VN_AS(argp->nextp(), Var)) {
+        bool found = false;
+        for (AstVarScope *vscp = ofuncp->scopep()->varsp(), *nextp; vscp; vscp = nextp) {
```

Comment:
Think this is O(n^2) on function variable count.  Record all arguments then look them up in one pass through scopep() varscopes?